lib/imgbundler/imgbundler.go

```diff
@@ -6,156 +6,202 @@ import (
 
-type resp struct {
-	srctxt string
-	data   string
-	err    error
+func BundleLocal(ctx context.Context, ms *xmain.State, in []byte) ([]byte, error) {
+	return bundle(ctx, ms, in, false)
 }
 
-func InlineLocal(ctx context.Context, ms *xmain.State, in []byte) ([]byte, error) {
-	return inline(ctx, ms, in, false)
+func BundleRemote(ctx context.Context, ms *xmain.State, in []byte) ([]byte, error) {
+	return bundle(ctx, ms, in, true)
 }
 
-func InlineRemote(ctx context.Context, ms *xmain.State, in []byte) ([]byte, error) {
-	return inline(ctx, ms, in, true)
+type repl struct {
+	from []byte
+	to   []byte
 }
 
-func inline(ctx context.Context, ms *xmain.State, svg []byte, isRemote bool) (_ []byte, err error) {
-	defer xdefer.Errorf(&err, "failed to bundle images")
-	imgs := imageRe.FindAllSubmatch(svg, -1)
+func bundle(ctx context.Context, ms *xmain.State, svg []byte, isRemote bool) (_ []byte, err error) {
+	if isRemote {
+		defer xdefer.Errorf(&err, "failed to bundle remote images")
+	} else {
+		defer xdefer.Errorf(&err, "failed to bundle local images")
+	}
+	imgs := imageRegex.FindAllSubmatch(svg, -1)
+	imgs = filterImageElements(imgs, isRemote)
+
+	ctx, cancel := context.WithTimeout(ctx, time.Minute*5)
+	defer cancel()
+
+	return runWorkers(ctx, ms, svg, imgs, isRemote)
+}
 
-	var filtered [][][]byte
+// filterImageElements finds all image elements in imgs that are eligible
+// for bundling in the current context.
+func filterImageElements(imgs [][][]byte, isRemote bool) [][][]byte {
+	imgs2 := imgs[:0]
 	for _, img := range imgs {
-		u, err := url.Parse(string(img[1]))
+		href := string(img[1])
+
+		// Skip already bundled images.
+		if strings.HasPrefix(href, "data:") {
+			continue
+		}
+
+		u, err := url.Parse(href)
 		isRemoteImg := err == nil && strings.HasPrefix(u.Scheme, "http")
+
 		if isRemoteImg == isRemote {
-			filtered = append(filtered, img)
+			imgs2 = append(imgs2, img)
 		}
 	}
+	return imgs2
+}
 
+func runWorkers(ctx context.Context, ms *xmain.State, svg []byte, imgs [][][]byte, isRemote bool) (_ []byte, err error) {
 	var wg sync.WaitGroup
-	respChan := make(chan resp)
+	replc := make(chan repl)
+
+	wg.Add(len(imgs))
+	go func() {
+		wg.Wait()
+		close(replc)
+	}()
+
 	// Limits the number of workers to 16.
 	sema := make(chan struct{}, 16)
 
-	ctx, cancel := context.WithTimeout(ctx, time.Minute*5)
-	defer cancel()
+	var errhrefsMu sync.Mutex
+	var errhrefs []string
 
-	wg.Add(len(filtered))
 	// Start workers as the sema allows.
 	go func() {
-		for _, img := range filtered {
+		for _, img := range imgs {
+			img := img
 			sema <- struct{}{}
-			go func(src, href string) {
+			go func() {
 				defer func() {
 					wg.Done()
 					<-sema
 				}()
 
-				var data string
-				var err error
-				if isRemote {
-					data, err = fetch(ctx, href)
-				} else {
-					data, err = read(href)
+				bundledImage, err := worker(ctx, img[1], isRemote)
+				if err != nil {
+					ms.Log.Error.Printf("failed to bundle %s: %v", img[0], err)
```

Comment:
u want to log `img[1]`, the href

<img width="801" alt="Screen Shot 2022-11-30 at 10 18 21 AM" src="https://user-images.githubusercontent.com/3120367/204877027-18592b47-a116-4f54-90e4-03815099408e.png">